scalanet/src/io/iohk/scalanet/peergroup/kademlia/KRouter.scala

```diff
@@ -3,190 +3,192 @@ package io.iohk.scalanet.peergroup.kademlia
-      addNodeRecord(nodeRecord)
-      Continue
-    } else { // the bucket is full
-      // ping the bucket's least recently seen node (i.e. the one at the head) to see what to do
-      val recordToPing = nodeRecords(bucket.head)
-      network
-        .ping(recordToPing, Ping(uuidSource(), config.nodeRecord))
-        .runToFuture
-        .flatMap { _ =>
-          // if it does respond, it is moved to the tail and the other node record discarded.
-          touchNodeRecord(recordToPing)
-          debug(s"Ping to ${recordToPing.id.toHex} in bucket $iBucket successful.")
-          info(
-            s"Moving ${recordToPing.id} to head of bucket $iBucket. Discarding (${nodeRecord.id.toHex}, $nodeRecord) as routing table candidate."
-          )
-          debug(s"iBucket($iBucket) = $bucket")
-          Continue
+  def add(nodeRecord: NodeRecord[A]): Task[Unit] = {
+    info(s"Handling potential addition of candidate (${nodeRecord.id.toHex}, $nodeRecord)")
+    for {
+      toPing <- routerState.modify { current =>
+        val (_, bucket) = current.kBuckets.getBucket(nodeRecord.id)
+        if (bucket.size < config.k) {
+          (current.addNodeRecord(nodeRecord), None)
+        } else {
+          // the bucket is full, not update it but ping least recently seen node (i.e. the one at the head) to see what to do
+          val nodeToPing = current.nodeRecords(bucket.head)
+          (current, Some(nodeToPing))
         }
-        .recover {
-          case _ =>
-            // if that node fails to respond, it is evicted from the bucket and the other node inserted (at the tail)
-            assert(bucket.size <= config.k)
-            removeNodeRecord(recordToPing)
-            addNodeRecord(nodeRecord)
-            debug(s"Ping to ${recordToPing.id.toHex} in bucket $iBucket failed.")
-            info(s"Replacing ${recordToPing.id.toHex} with new entry (${nodeRecord.id.toHex}, $nodeRecord).")
-            Continue
-        }
-    }
-  }
-
-  private def addNodeRecord(nodeRecord: NodeRecord[A]): Unit = {
-    nodeRecordIndex = NodeRecordIndex(kBuckets.add(nodeRecord.id), nodeRecords + (nodeRecord.id -> nodeRecord))
-  }
-
-  private def removeNodeRecord(nodeRecord: NodeRecord[A]): Unit = {
-    nodeRecordIndex = NodeRecordIndex(kBuckets.remove(nodeRecord.id), nodeRecords - nodeRecord.id)
-  }
-
-  private def touchNodeRecord(nodeRecord: NodeRecord[A]): Unit = {
-    nodeRecordIndex = NodeRecordIndex(kBuckets.touch(nodeRecord.id), nodeRecords)
-  }
-
-  private def getRemotely(key: BitVector): Future[NodeRecord[A]] = {
-    lookup(key).flatMap(_ => getLocally(key))
-  }
-
-  private def getLocally(key: BitVector): Future[NodeRecord[A]] = {
-    toFuture(
-      for {
-        nodeId <- kBuckets.closestNodes(key, 1).find(_ == key)
-        record <- nodeRecords.get(nodeId)
-      } yield record,
-      new Exception(s"Target node id ${key.toHex} not loaded into kBuckets.")
-    )
+      }
+      result <- if (toPing.isEmpty) {
```

Comment:
This could be a `ping` function or something like that to avoid having nested for comprehensions